app/src/main/java/org/thoughtcrime/securesms/messages/IncomingMessageObserver.java

```diff
@@ -0,0 +1,283 @@
+      public void onBackground() {
+        onAppBackgrounded();
+      }
+    });
+
+    connectionReceiver = new BroadcastReceiver() {
+      @Override
+      public void onReceive(Context context, Intent intent) {
+        synchronized (IncomingMessageObserver.this) {
+          if (!NetworkConstraint.isMet(context)) {
+            Log.w(TAG, "Lost network connection. Shutting down our websocket connections and resetting the drained state.");
+            networkDrained    = false;
+            decryptionDrained = false;
+            disconnect();
+          }
+          IncomingMessageObserver.this.notifyAll();
+        }
+      }
+    };
+
+    context.registerReceiver(connectionReceiver, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+  }
+
+  public synchronized void notifyRegistrationChanged() {
+    notifyAll();
+  }
+
+  public synchronized void addDecryptionDrainedListener(@NonNull Runnable listener) {
+    decryptionDrainedListeners.add(listener);
+    if (decryptionDrained) {
+      listener.run();
+    }
+  }
+
+  public boolean isDecryptionDrained() {
+    return decryptionDrained || networkAccess.isCensored(context);
+  }
+
+  public void notifyDecryptionsDrained() {
+    List<Runnable> listenersToTrigger = new ArrayList<>(decryptionDrainedListeners.size());
+
+    synchronized (this) {
+      if (networkDrained && !decryptionDrained) {
+        Log.i(TAG, "Decryptions newly drained.");
+        decryptionDrained = true;
+        listenersToTrigger.addAll(decryptionDrainedListeners);
+      }
+    }
+
+    for (Runnable listener : listenersToTrigger) {
+      listener.run();
+    }
+  }
+
+  private synchronized void onAppForegrounded() {
+    appVisible = true;
+    notifyAll();
+  }
+
+  private synchronized void onAppBackgrounded() {
+    appVisible = false;
+    notifyAll();
+  }
+
+  private synchronized boolean isConnectionNecessary() {
+    boolean registered          = TextSecurePreferences.isPushRegistered(context);
+    boolean websocketRegistered = TextSecurePreferences.isWebsocketRegistered(context);
+    boolean isGcmDisabled       = TextSecurePreferences.isFcmDisabled(context);
+    boolean hasNetwork          = NetworkConstraint.isMet(context);
+    boolean hasProxy            = SignalStore.proxy().isProxyEnabled();
+
+    Log.d(TAG, String.format("Network: %s, Foreground: %s, FCM: %s, Censored: %s, Registered: %s, Websocket Registered: %s, Proxy: %s",
+                             hasNetwork, appVisible, !isGcmDisabled, networkAccess.isCensored(context), registered, websocketRegistered, hasProxy));
+
+    return registered                    &&
+           websocketRegistered           &&
+           (appVisible || isGcmDisabled) &&
+           hasNetwork                    &&
+           !networkAccess.isCensored(context);
+  }
+
+  private synchronized void waitForConnectionNecessary() {
+    try {
+      while (!isConnectionNecessary()) wait();
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)